cluster/gce/upgrade.sh

```diff
@@ -111,13 +113,57 @@ function prepare-upgrade() {
   fi
 }
 
+# Reads kube-env metadata from master and extracts value from provided key.
+#
+# Assumed vars:
+#   MASTER_NAME
+#   ZONE
+#
+# Args:
+# $1 env key to use
+function get-env-val() {
+  gcloud compute ssh --zone ${ZONE} ${MASTER_NAME} --command \
+    "curl --fail --silent -H 'Metadata-Flavor: Google' \
+      'http://metadata/computeMetadata/v1/instance/attributes/kube-env'" 2>/dev/null \
+    | grep ${1} | cut -d : -f 2 | cut -d \' -f 2
+}
+
+# $1 veresion
 function upgrade-nodes() {
-  echo "== Upgrading nodes to ${SERVER_BINARY_TAR_URL}. =="
+  local version=${1}
+  local sanitized_version=$(echo ${version} | sed s/"\."/-/g)
+  echo "== Upgrading nodes to ${version}. =="
 
   detect-minion-names
-  get-password
-  set-master-htpasswd
-  kube-update-nodes upgrade
+
+  # TODO(mbforbes): Refactor setting scope flags.
+  local -a scope_flags=()
+  if (( "${#MINION_SCOPES[@]}" > 0 )); then
+    scope_flags=("--scopes" "${MINION_SCOPES[@]}")
+  else
+    scope_flags=("--no-scopes")
+  fi
+
+  # Get required node tokens.
+  KUBELET_TOKEN=$(get-env-val "KUBELET_TOKEN")
+  KUBE_PROXY_TOKEN=$(get-env-val "KUBE_PROXY_TOKEN")
+
+  # TODO(mbforbes): How do we ensure kube-env is written in a ${version}-
+  #                 compatible way?
+  write-node-env
+  # TODO(mbforbes): Get configure-vm script from ${version}. (Must plumb this
+  #                 through all create-node-instance-template implementations).
+  create-node-instance-template ${sanitized_version}
+
+  # Do the actual upgrade.
+  gcloud preview rolling-updates start \
+      --group "${NODE_INSTANCE_PREFIX}-group" \
+      --max-num-concurrent-instances 1 \
+      --max-num-failed-instances 1 \
```

Comment:
Agh, I read the docs backwards. Thought it meant max num that can fail before the whole thing is failed, not _without_ failing the whole thing. Switched to 0. (I'm going to push without squashing so these comments don't go away.)